### DISCUSSION TITLE
Sending HTTP GET request via CLI and TCP

### DISCUSSION THREAD

**Discussion by ghost** [openthread/openthread#7904] [Q&A] [answered]
I'm using an nRF52840dk and proved that I can successfully send TCP messages (ex. "test123" or "test\ with\ spaces") to an nRF52840dongle on the same OpenThread network. I'm now attempting to send a HTTP GET request via the CLI tool with TCP enabled. I can ping 74.125.142.104 as seen here
```
uart:~$ ot ping 64:ff9b::4a7d:8e68
16 bytes from 64:ff9b:0:0:0:0:4a7d:8e68: icmp_seq=1 hlim=104 time=79ms
1 packets transmitted, 1 packets received. Packet loss = 0.0%. Round-trip min/avg/max = 79/79.0/79 ms.
Done
```
However, I'm having trouble with the syntax for a multi-line HTTP GET request via the CLI tool.
```
uart:~$ ot tcp init
Done
uart:~$ ot tcp connect 64:ff9b::4a7d:8e68 443
Done
TCP: Connection established
uart:~$ ot tcp send "GET\ /\ HTTP/1.1\r\nhost:\ www.google.com\r\n"
Done
TCP: Reached end of stream
TCP: Connection reset
uart:~$ ot tcp send "GET\ /\ HTTP/1.1\r\nhost:\ www.google.com\r\n"
Error 1: Failed
uart:~$
```
Am I doing this incorrectly or is the CLI tool not designed for this type of multi line use case?

EDIT: Reformatted code snippets
EDIT2: I should note that I'm working my way toward connecting with Azure IoT Hub, similar to Discussion Topic #7784. Also, I've noticed an increase in interest regarding the use of OpenThread, TCP, and Azure IoT Hub across various forums in the last ~2-3 months.

**Comment by jwhui** [comment, marked as answer]
I would suggest taking a look at the wireshark trace. I believe the TCP stream is not sending the exact bytes that you are expecting and the HTTP server is responding with "Bad Request".

In particular, OpenThread's current [command line parser implementation](https://github.com/openthread/openthread/blob/e538aad0951c87cfdcaf678a5c64a0acc53b5b58/src/core/utils/parse_cmdline.cpp#L88) does not transform escaped characters like `\r` and `\n` into the appropriate ASCII characters.

**Reply by ghost** [reply]
Thank you for the help. I'll mark as closed since the escaped characters are certainly something I had incorrect. 

For what it's worth, my eventual goal was to validate connecting an OpenThread end device to an Azure IoT Hub server via TCP, and I have done this through other means (a combination of the Nordic azure_iot_hub and azure_fota samples). Although I still get an mqtt_connect error (-22, invalid argument), I see that the end device and Azure are at least trading TCP messages.

![image](https://user-images.githubusercontent.com/104365393/180294310-ce77fbff-0b39-42d9-be8a-447984bdbc09.png)
